### PR TITLE
Update README for 1.2 preview, with associated fixes

### PR DESCRIPTION
**Pull request by tcard**
Also, I was anxious about fixing README without actually checking that the fixes are correct, so I added script to generate a Go test from it.

[Rendered README](https://github.com/ably/ably-go/blob/feature/1.2-preview-readme/README.md)

### CODE REVIEW

File: README.md
Comment (tcard) [inline README.md:264]:
```suggestion
- Realtime connection failure handling is partially implemented
```

File: README.md
Comment (lmars) [inline README.md:218]: This is the first time I've seen how this API is used, so perhaps this isn't the correct place to discuss, but I'll put it here anyway since I've seen it.

The usage feels a little awkward, I wonder if something like the following would be better:

```go
params := &ably.PaginateParams{
	Limit: 10,
}
err := channel.Presence.Get(params, func(page *ably.PaginatedResult) bool {
	// iterate page.PresenceMessages()
	// return false to stop iterating
})
```

and similarly for the other paginated operations.

File: README.md
Comment (lmars) [inline README.md:241]: Are these no longer relevant?

File: README.md
Comment (lmars) [inline README.md:264]: This is surprising, what is left to implement? I think this is one of the main features we're aiming to give to customers in this release.

File: README.md
Comment (lmars) [inline README.md:269]: The examples use `1.1.x`, but this says semver only started with `1.2.x`, so should the examples reference `1.2.x`, or was semver already in use for `1.1.x`?

File: README.md
Comment (tcard) [reply README.md:17]: > Should that also say 1.2?

True. Fixed at e9b274a.



> Also, why does the library need to be upgraded incrementally? Are we saying users have to upgrade to 1.1, then to 1.2?

As I understand it, this means that version 1.1 (now, 1.2) doesn't implement the whole 1.1 spec. While we haven't really decided this, I'd expect it will be the same case for 1.2. Version 1.2.0 can be released as soon as we have a partially spec-compliant version onto which 1.2 features can be added in a backwards-compatible way in further minor releases.

File: README.md
Comment (tcard) [reply README.md]: Moved and reworded at 03b766c.

File: README.md
Comment (tcard) [reply README.md:184]: Note this API isn't covered by the 1.2 preview subset and is likely to change. I just fixed it here (previous version returned an error on end of iteration, so this example would always panic).

I'm moving this discussion at https://github.com/ably/ably-go/issues/197#issuecomment-718446264 if that's OK.

File: README.md
Comment (tcard) [reply README.md:241]: Those weren't really usage examples. They lack arguments and don't use the return values.

Also, they arguably aren't relevant enough for most Go users, who are likely to prefer key auth given Go runs (in the overwhelming majority of cases) on private servers. So they're beyond the scope of an introductory README.

File: README.md
Comment (tcard) [reply README.md]: Once #172 is merged this line can be removed, but the API preview release that will result from _this_ PR won't have that (as things stand now).

File: README.md
Comment (tcard) [reply README.md:260]: Using 1.2.0 now as an example at df19c8b.

File: README.md
Comment (tcard) [reply README.md:22]: I'm not sure we should overwhelm the README with error paths. For a full reference, users can check the Ably docs and the Go package docs.

File: README.md
Comment (tcard) [reply README.md]: Done at ad3bc49.

File: README.md
Comment (tcard) [reply README.md:77]: Added at ad3bc49.

File: README.md
Comment (tcard) [reply README.md:173]: Added at ad3bc49.

File: README.md
Comment (lmars) [reply README.md]: > Version 1.2.0 can be released as soon as we have a partially spec-compliant version onto which 1.2 features can be added in a backwards-compatible way in further minor releases.

Ok, I agree with that, but I'm still confused by the sentence in the README, since it seems to suggest that the user needs to do some form of incremental upgrade (e.g. upgrade to 1.2.1, then 1.2.2, then 1.2.3 etc.), but I don't believe that to be the case.

Perhaps the section would be clearer as something like:

> This library targets the Ably 1.2 client library specification, although as of release 1.2.0 the following are not implemented and will be covered in future 1.2.x releases:
>
> [list of unimplemented things]
>
> If any of these unimplemented features are a high priority for your use-case then please contact Ably customer support. Pull Requests are also welcomed.

File: README.md
Comment (lmars) [reply README.md:22]: > I'm not sure we should overwhelm the README with error paths

This is a pretty important error path for users to be aware of though, no?

> For a full reference, users can check the Ably docs and the Go package docs.

Ok, can we at least put a link here to where those exist for handling connection errors?

File: README.md
Comment (lmars) [reply README.md:209]: That's fine, thanks 👍

File: README.md
Comment (lmars) [inline README.md:128]: This should use `fmt.Printf`

File: README.md
Comment (tcard) [reply README.md]: See 1e4ff57.

File: README.md
Comment (tcard) [reply README.md]: Done at f2b682779470cb2b98833d3f702026984efdfe27.

File: README.md
Comment (tcard) [reply README.md:23]: See cbdaa10.

Those examples are how things _should_ be, but they won't work at least until #172 is merged, and they probably depend on [RTL6 too](https://github.com/ably/ably-go/pull/172#issuecomment-718666302) (which I'm working on now). But they showcase how users are likely to encounter a connection error: by trying to use the connection for publishing, subscribing or some other operation.

I can revert that commit for now so that this PR is mergeable, or let the PR blocked until those issues are fixed and then rebase.

Comment (lmars) [reply README.md:23]: > See cbdaa10.

Great thanks 👍 

> I can revert that commit for now so that this PR is mergeable, or let the PR blocked until those issues are fixed and then rebase.

I think we can merge this and just focus on merging the other PR and related fixes before we share this updated README with people.

**Review by lmars — APPROVED** [review]

**Pull request merged**
